### ISSUE TITLE
ECJ + Bump lombok from 1.18.16 to 1.18.18 = Unknown source: java.lang.IllegalArgumentException: argument type mismatch

### ISSUE THREAD

**Issue by pzygielo** [rzwitserloot/lombok#2729]
**Describe the bug**
Bumping lombok from 1.18.16 to 1.18.18 on project compiled with maven using ecj causes following
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project will: Compilation failure
Error:  Unknown source: java.lang.IllegalArgumentException: argument type mismatch
```

Upgraded lombok works fine in the same project when compiled with javac.

**To Reproduce**
https://github.com/pzrep/lombok-2729
with upgrade failed in https://github.com/pzrep/lombok-2729/pull/5

**Expected behavior**
Build completed with success.

**Version info (please complete the following information):**
 - 1.18.18
 - maven + plexus-compiler-eclipse + org.eclipse.jdt:ecj:3.24.0

**Comment by pzygielo** [comment]
I think it's something between v1.18.16 (737dc6e4) and edge (32172755) - as build with
```xml
<snapshot>
  <timestamp>20201204.222028</timestamp>
  <buildNumber>1</buildNumber>
</snapshot>
```
yields the same result `java.lang.IllegalArgumentException: argument type mismatch`.

**Comment by Rawi01** [comment]
I think the changes in 28fbff467b3f3f3f5a4128313de02c9b1b2cb41d cause this problem.

**Comment by JohannesBeranek** [comment]
Same issue on Intellij IDEA 2020.3.2 using javac. Unfortunately that IDEA version only works with newer lombok version (due to bundled lombok plugin), but at the same time our programs won't run anymore due to this issue (and various other issues in 1.18.16)

**Comment by Rawi01** [comment]
@JohannesBeranek This is an `ecj` bug and I'm pretty sure that it works if you use `javac`, at least I'm not aware of general `javac` issues. If there is a `javac` related problem please create a new issue or add a comment or +1 on an existing issue.

You cannot use older lombok versions in IDEA because they changed how they process annotations and the older lombok versions do not know how to handle that.

**Comment by JohannesBeranek** [comment]
@Rawi01 thanks, created a ticket here https://github.com/rzwitserloot/lombok/issues/2738 ; exception thrown is the same

**Comment by pzygielo** [comment]
Update about just released edge (20210319.040806-1.pom) - it works with ecj pretty fine! :+1: 

Thank you.
